app/controllers/concerns/find_common.rb

```diff
@@ -10,23 +10,39 @@ def find_resource
   end
 
   def resource_name
-    controller_name.singularize
+    resource_name_for(controller_name)
   end
 
   def resource_class
-    @resource_class ||= resource_name.classify.constantize
+    @resource_class ||= resource_class_for(resource_name)
   end
 
   def resource_scope(options = {})
-    @resource_scope ||= begin
-      options[:controller] ||= controller_name
-      options[:permission] ||= "#{action_permission}_#{options[:controller]}"
-      scope = resource_class.scoped
-      if resource_class.respond_to?(:authorized)
-        scope.authorized(options[:permission], resource_class)
-      else
-        scope
-      end
+    @resource_scope ||= scope_for(resource_class, options)
+  end
+
+  def scope_for(resource, options = {})
+    controller = options.delete(:controller){ controller_name }
+    #don't call the #action_permission method here, we are not sure if the resource is authorized at this point
+    permission = options.delete(:permission)
+
+    if resource.respond_to?(:authorized)
+      permission ||= "#{action_permission}_#{controller}"
+      resource = resource.authorized(permission, resource)
+    end
+
+    resource.where(options)
+  end
+
+  def resource_name_for(resource)
+    resource.singularize
+  end
+
+  def resource_class_for(resource)
+    begin
+      return resource.classify.constantize
+    rescue NameError
```

Comment:
The concept of reflecting on foreign keys is good, but what about polymorphic relations like taxonomies? I mean in the query you will have `organization_id` but there is no such foreign key, the same goes for smart proxies.
`resource_class` uses a special case of `resource_class_for` - the one with `resource_name` which is derived from `controller_name`. If it's not good enough I can add an explicit check to `resource_class` that will throw an exception in case the value returned empty.
If the solution is acceptable, I'll just rebase the code.